python/shotgun_model/shotgun_query_model.py

```diff
@@ -0,0 +1,947 @@
+        a default thumbnail.  In the current implementation, thumbnails are not
+        cached in the same way as the rest of the model data, meaning that this
+        method is executed each time an item is constructed, regardless of if
+        it came from an asynchronous shotgun query or a cache fetch.
+
+        The purpose of this method is that you can subclass it if you want to
+        ensure that items have an associated thumbnail directly when they are
+        first created.
+
+        Later on in the data load cycle, if the model was instantiated with the
+        `download_thumbs` parameter set to True, the standard Shotgun ``image``
+        field thumbnail will be automatically downloaded for all items (or
+        picked up from local cache if possible). When these real thumbnails
+        arrive, the meth:`_populate_thumbnail()` method will be called.
+
+        :param item: :class:`~PySide.QtGui.QStandardItem` that is about to be
+            added to the model.  This has been primed with the standard
+            settings that the ShotgunModel handles.
+        """
+        # the default implementation does nothing
+        pass
+
+    def _populate_item(self, item, sg_data):
+        """
+        Whenever an item is downloaded from Shotgun, this method is called. It
+        allows subclasses to intercept the construction of a
+        :class:`~PySide.QtGui.QStandardItem` and add additional metadata or
+        make other changes that may be useful. Nothing needs to be returned.
+
+        This method is called before the item is added into the model tree. At
+        the point when the item is added into the tree, various signals will
+        fire, informing views and proxy models that a new item has been added.
+        This methods allows a subclassing object to add custom data prior to
+        this.
+
+        .. note:: When an item is fetched from the cache, this method is *not*
+            called, it will only be called when shotgun data initially arrives
+            from a Shotgun API query.
+
+        :param item: :class:`~PySide.QtGui.QStandardItem` that is about to be
+            added to the model.
+
+        :param sg_data: Shotgun data dictionary that was received from Shotgun.
+        """
+        # default implementation does nothing
+        pass
+
+    def _set_tooltip(self, item, data):
+        """
+        Sets the tooltip for the supplied item.
+
+        Called when an item is created.
+
+        :param item: Shotgun model item that requires a tooltip.
+        :param data: Dictionary of the SG data associated with the model.
+        """
+        # the default implementation does not set a tooltip
+        pass
+
+    ############################################################################
+    # protected convenience methods. these methods can be used by subclasses
+    # to manipulate and manage data returned from Shotgun.
+
+    def _do_depth_first_tree_deletion(self, node):
```

Comment:
Yep. I think at one point this was used in the subclasses and then I managed to move it down here. I've made it private.

fyi, that's why much of this ended up this way. I made things protected in order for them to be used by the subclasses. Then as I unraveled a bit, I was able to move more calls back to the base class and I just didn't make them private again. So this is good stuff.